engine/execution/computation/computer/uploader/uploader_test.go

```diff
@@ -95,35 +99,75 @@ func Test_AsyncUploader(t *testing.T) {
 
 		callCount := 0
+		t.Log("test started grID:", string(bytes.Fields(debug.Stack())[1]))
 
-		wg := sync.WaitGroup{}
-		wg.Add(1)
+		// this wait group ensures that async uploader has a chance to start the upload before component is shut down
+		// otherwise, there's a race condition that can happen where the component can shut down before the async uploader
+		// has a chance to start the upload
+		wgUploadStarted := sync.WaitGroup{}
+		wgUploadStarted.Add(1)
+
+		// this wait group ensures that async uploader won't send an error (to test if retry will kick in) until
+		// the component has initiated shutting down (which should stop retry from working)
+		wgShutdownStarted := sync.WaitGroup{}
+		wgShutdownStarted.Add(1)
+		t.Log("added 1 to wait group grID:", string(bytes.Fields(debug.Stack())[1]))
 
 		uploader := &DummyUploader{
 			f: func() error {
-				defer func() {
+				t.Log("DummyUploader func() - about to call wgUploadStarted.Done() grID:", string(bytes.Fields(debug.Stack())[1]))
+				// signal to main goroutine that upload started, so it can initiate shutting down component
+				wgUploadStarted.Done()
+
+				t.Log("DummyUpload func() waiting for component shutdown to start grID:", string(bytes.Fields(debug.Stack())[1]))
+				wgShutdownStarted.Wait()
+				t.Log("DummyUploader func() component shutdown started, about to return error grID:", string(bytes.Fields(debug.Stack())[1]))
+
+				// force an upload error to test that upload is never retried (because component is shut down)
+				// normally, we would see retry mechanism kick in and the callCount would be > 1
+				// but since component has started shutting down, we expect callCount to be 1
+				if callCount < 5 {
+					t.Logf("DummyUploader func() incrementing callCount=%d grID: %s", callCount, string(bytes.Fields(debug.Stack())[1]))
 					callCount++
-				}()
-				wg.Wait()
-				return fmt.Errorf("this should return only once")
+					t.Logf("DummyUploader func() about to return error callCount=%d grID: %s", callCount, string(bytes.Fields(debug.Stack())[1]))
+					return fmt.Errorf("this should return only once")
+				}
+				return nil
 			},
 		}
-
+		t.Log("about to create NewAsyncUploader grID:", string(bytes.Fields(debug.Stack())[1]))
 		async := NewAsyncUploader(uploader, 1*time.Nanosecond, 5, zerolog.Nop(), &metrics.NoopCollector{})
-
+		t.Log("about to call async.Upload() grID:", string(bytes.Fields(debug.Stack())[1]))
 		err := async.Upload(computationResult) // doesn't matter what we upload
 		require.NoError(t, err)
 
+		// stop component and check that it's fully stopped
+		t.Log("about to close async uploader grID:", string(bytes.Fields(debug.Stack())[1]))
+
+		// wait until upload has started before shutting down the component
+		wgUploadStarted.Wait()
+
+		// stop component and check that it's fully stopped
+		t.Log("about to initiate shutdown grID: ", string(bytes.Fields(debug.Stack())[1]))
 		c := async.Done()
-		wg.Done()
-		<-c
```

Comment:
Don't we need to read from a channel anymore?

Reply:
this is done with the helper method on 168: `unittest2.RequireCloseBefore(t, c, 1*time.Second, "async uploader not closed in time")` which has a timeout in case reading form the channel results in a deadlock